BigQuery Interpreter for Apazhe Zeppelin[ZEPPELIN-1153]

babupe opened the pull request:
### What is this PR for?
Google BigQuery is a popular no-ops datawarehouse. This commit will enable Apache Zeppelin users to perform BI and Analytics on their datasets in BigQuery.

### What type of PR is it?
Feature

### Todos
* Make bigquery interpreter appear in the interpreters section in the UI
* Build SQL completion 
* Authorization of non-gcp 

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-1153

### How should this be tested?
copy conf/zeppelin-site.xml.template to conf/zeppelin-site.xml
Add org.apache.zeppelin.bigquery.bigQueryInterpreter to property zeppelin.interpreters in zeppelin-site.xml
Start Zeppelin
Add BigQuery Interpreter with your project ID
Create new note with %bsql.sql and run your SQL against public datasets in bigquery.

### Screenshots (if appropriate)
![screenshot from 2016-07-12 14 27 30](https://cloud.githubusercontent.com/assets/4242273/16785302/31b104e2-4842-11e6-87c0-b79763dd85c0.png)

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

bzz commented on bigquery/src/main/java/org/apache/zeppelin/bigquery/bigQueryInterpreter.java:
```diff
@@ -0,0 +1,322 @@
+    String projId = getProperty(PROJECT_ID);
+    long wTime = Long.parseLong(getProperty(WAIT_TIME));
+    Iterator<GetQueryResultsResponse> pages = run(sql, projId, wTime);
+    while (pages.hasNext()) {
```
Usually we want to handle errors here as well (I.e bad SQL syntax or network issues) and set paragraph error status to Error by returning `InterpreterResult(Code.ERROR)` in such cases.

I.e by looking at the code below, it seems that `pages` can be `null` in case of connectivity issues, resulting in the NPE instead.

bzz commented:
Great contribution, thank you!

Couple of questions on top of few one already raised above:
 - will you be willing to add some tests as well? Looking at other interpreters impl like #1164 one can see how to test I.e syntax errors and other conditions, etc
 - would it be hard or do you have plans to implement `cancel` and `getProgress` as part of this contribution? They are very usefull in case of big queries and improve user experience a lot on UI
 - in current code, you use static interpreter registration mechanism, which is now [deprecated](https://issues.apache.org/jira/browse/ZEPPELIN-804) and all interpreters are in the process of migration. While we are here, in order not to create one more migration sub-tesk for later, could you please switch to `interpreter-settings.json` one instead? Its a dime change, here is an example of how it is done for another interpreter #1063 
 - every interpreter and it's supported/unsupported features and cobfiguration options are usually documented under `/docs/` which becomes part of the [website](http://zeppelin.apache.org/docs/0.6.0/development/howtocontributewebsite.html). Would you care to document this one as well?
 - for every new library introduced by this module, which Jar will end up in the release convenience binary (all, with default scope) we need to have a record in `https://github.com/apache/zeppelin/blob/master/zeppelin-distribution/src/bin_license/LICENSE` specifying it's license

Feel free to ask here, if something is not clearly or in case you have any further questions!

AhyoungRyu commented on bigquery/src/main/java/org/apache/zeppelin/bigquery/bigQueryInterpreter.java:
```diff
@@ -0,0 +1,322 @@
+ * </p>
+ * 
+ */
+
```
The above comments are better for `bigquery.md` as @bzz  said in [https://github.com/apache/zeppelin/pull/1170#issuecomment-232215164](https://github.com/apache/zeppelin/pull/1170#issuecomment-232215164). You can refer to many existed docs [in here](https://github.com/apache/zeppelin/tree/master/docs/interpreter). It would be helpful to users who want to try BigQuery in Zeppelin.

babupe commented:
Thanks! I will work on incorporating the feedback.

AhyoungRyu commented on bigquery/src/main/resources/interpreter-setting.json:
```diff
@@ -0,0 +1,27 @@
+      }
+    }
+  }
+]
```
You can add the above property list to `bigquery.md` as a configuration table like [this](https://zeppelin.apache.org/docs/0.7.0-SNAPSHOT/interpreter/alluxio.html#configuration).

AhyoungRyu commented:
@babupe Also you need to add
```
<li><a href="{{BASE_PATH}}/interpreter/bigquery.html">BigQuery</a></li>
```

this line to [here](https://github.com/apache/zeppelin/blob/master/docs/_includes/themes/zeppelin/_navigation.html#L51). Then you can find `bigquery.md` in the dropdown menu on docs site.

babupe commented:
Hi,

1. I have added unit tests, but the tests will work with a valid google project id on google compute, else it will fail. This is because the test has to run against the bigquery service which does not have a local build.
2. I have added a server side cancel operation. However I could not find a viable method to report progress. Will implement this if this becomes available as part of big query API in the future.
3. Implemented dynamic interpreter
4. Added BigQuery API license to the license file
5. Docs updated with BigQuery information.
6. Exceptions implemented at appropriate places.

Thanks!

AhyoungRyu commented on docs/_includes/themes/zeppelin/_navigation.html:
```diff
@@ -66,6 +66,7 @@
                 <li><a href="{{BASE_PATH}}/interpreter/scalding.html">Scalding</a></li>
                 <li><a href="{{BASE_PATH}}/interpreter/shell.html">Shell</a></li>
                 <li><a href="{{BASE_PATH}}/interpreter/spark.html">Spark</a></li>
+                <li><a href="{{BASE_PATH}}/interpreter/bigquery.html">BigQuery</a></li>
```
As you can see, the lists are in alphabetical order. You need to put `bigquery.html` between `alluxio.html` and `cassandra.html`.

babupe replied:
Fixed this

bzz commented:
@babupe thank you for taking care. @AhyoungRyu thank you for review. 

Let me do another pass on it and get back to you.

bzz commented on bigquery/src/main/java/org/apache/zeppelin/bigquery/bigQueryInterpreter.java:
```diff
@@ -0,0 +1,340 @@
+ * Copyright 2016 Google Inc.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
```
I would strongly prefer keeping original text of the license here: `See the NOTICE file distributed with this work for additional information regarding copyright ownership.` 

So copyright notice need to be moved to root NOTICE file and refer to `./bigquery/` module.

babupe replied:
So, the google copyright should be completely removed here?
Where should I include the Google copyright then? Can you share an example?

bzz commented on bigquery/src/main/java/org/apache/zeppelin/bigquery/bigQueryInterpreter.java:
```diff
@@ -0,0 +1,340 @@
+      }
+      return new InterpreterResult(Code.SUCCESS, finalmessage.toString());
+    }
+    catch ( NullPointerException ex ) {
```
a nit: looks like indentation is a bit off here.
To be consistent with the rest of the project and [styleguide](https://google.github.io/styleguide/javaguide.html#s6.2-caught-exceptions) here and in other places `catch` needs to be on prev. line

babupe replied:
Got it. Fixed catch statements all over code to be on same line are ending "}"

babupe replied:
We dont want to throw an error when there is no next page, but only when the page pointer is null.

bzz commented on bigquery/src/main/java/org/apache/zeppelin/bigquery/bigQueryInterpreter.java:
```diff
@@ -0,0 +1,340 @@
+    }
+    catch (IOException e) {
+      e.printStackTrace();
+      return null;
```
We use Log4j for logging in Zeppelin, it would be better to use it and replace `printStackTrace()` with `log.error()`

babupe replied:
I have removed it since the null is handled in the calling function.

bzz commented on bigquery/pom.xml:
```diff
@@ -0,0 +1,192 @@
+
+    <dependency>
+      <groupId>jline</groupId>
+      <artifactId>jline</artifactId>
```
I'm a bit confused with dependency list - could you please point me where you use this library?

babupe replied:
Yeah, jline is not required. Must have been left over from a copy.
Removed it now

bzz commented on bigquery/src/main/java/org/apache/zeppelin/bigquery/bigQueryInterpreter.java:
```diff
@@ -0,0 +1,340 @@
+          if (response.containsKey("pageToken")) {
+            request = request.set("pageToken", response.get("pageToken"));
+          } 
+          else {
```
indentation looks a bit off here

babupe replied:
Fixed.

bzz commented on bigquery/src/main/java/org/apache/zeppelin/bigquery/bigQueryInterpreter.java:
```diff
@@ -0,0 +1,340 @@
+      return msg.toString();
+    }
+    catch ( NullPointerException ex ) {
+      throw new NullPointerException("SQL Execution returned an error!");
```
Allthough it's not obvious what is the error condition here - it's alwasy better to keep the cause:
```java
throw new NullPointerException("SQL Execution returned an error!", ex);
```

bzz commented on bigquery/src/main/java/org/apache/zeppelin/bigquery/bigQueryInterpreter.java:
```diff
@@ -0,0 +1,340 @@
+ */
+
+
+public class bigQueryInterpreter extends Interpreter {
```
Looks like a typo here, but it violates Java [class naming convention](https://google.github.io/styleguide/javaguide.html#s5.2.2-class-names)

babupe replied:
Fixed on all classes

bzz commented on docs/interpreter/bigquery.md:
````diff
@@ -0,0 +1,95 @@
+
+For Example, SQL to query for top 10 departure delays across airports using the flights public dataset
+
+```bash
````
Is that on purpose, so `sql` does not highlight the syntax?

babupe replied:
yes.

bzz commented:
@babupe great job, I have highlighted few issues inline in code above.

I would really appreciate if, for the sake of maintainability, you could please add `./bigquery/README.md` with some _technical_ information on interpreter implementation i.e

 - point to the sources of current BigQuery library that is beeing used

  ```
com.google.apis:google-api-services-bigquery:v2-rev265-1.21.0
  ```
  I had no luck locating it by searching online :\

 - small rationale why it was used or how it is compared to [GoogleCloudPlatform/gcloud-java](https://github.com/GoogleCloudPlatform/gcloud-java/)

 - mention authentication mechanism used

babupe commented:
https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/TESTING.md#testing-code-that-uses-bigquery
There is no emulator for BigQuery today.

babupe commented:
Please clarify regarding the LICENSE file.
Also, if you have any other feedback, would be happy to take a look.

AhyoungRyu commented:
+1 for technical docs `bigquery/README.md` that @bzz suggested and it looks great that you added! 
How about adding below one sentence in the end of `bigquery.md` so that users can know the existence of this file?

```
## Technical description

For in-depth technical details on current implementation please refer to [bigquery/README.md](https://github.com/apache/zeppelin/blob/master/bigquery/README.md).
```

bzz commented on bigquery/README.md:
```diff
@@ -0,0 +1,96 @@
+The Interpreter opens a connection with the BigQuery Service using the supplied Google project ID and the compute environment variables.
+
+# Google BigQuery API Javadoc
+[API Javadocs](https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/java/latest/)
```
This is JavaDoc for the artefact 

```
<groupId>com.google.apis</groupId>
<artifactId>google-api-services-bigquery</artifactId>
<version>v2-rev265-1.21.0</version>
```
right?

AFAIK it's an open-source library, so would you be so kind to add a link here to it's source code please? This could help future maintainers to keep up with changes, etc.

babupe replied:
Got it. These packages are licensed under Apache 2.0. I have asked around to see if the code is publicly available.

bzz replied:
Any updates on this one?

bzz commented on bigquery/pom.xml:
```diff
@@ -0,0 +1,186 @@
+  ~ Licensed to the Apache Software Foundation (ASF) under one or more
+  ~ contributor license agreements.  See the NOTICE file distributed with
+  ~ this work for additional information regarding copyright ownership.
+  ~ The ASF licenses this file to You under the Apache License, Version 2.0
```
This is the standard license hearer that is used in Apache Zeppelin project.
Could you please make sure that it's the same for all new files, added in this PR?

bzz commented:
Thanks @babupe I think we are almost there! Did another pass and add few last things above.

One more thing I noticed is - right now this PR fails the project's CI build and we heavily rely on this CI automation in the overall progress of the project. So we need to make it green again, in order to be able to merge this. 

From [the logs](https://s3.amazonaws.com/archive.travis-ci.org/jobs/145380794/log.txt) it looks like tests are failing

```
[INFO] Zeppelin: BigQuery interpreter ..................... FAILURE [  2.834 s]
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:testCompile (default-testCompile) on project zeppelin-bigquery: Compilation failure: Compilation failure:
[ERROR] /home/travis/build/apache/zeppelin/bigquery/src/test/java/org/apache/zeppelin/bigquery/BigQueryInterpreterTest.java:[41,41] package com.google.cloud.bigquery.testing does not exist
[ERROR] /home/travis/build/apache/zeppelin/bigquery/src/test/java/org/apache/zeppelin/bigquery/BigQueryInterpreterTest.java:[41,41] package com.google.cloud.bigquery.testing does not exist
[ERROR] -> [Help 1]
[ERROR] 
```

I tried running `mvn test -pl bigquery` on my machine and got 
```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/..../zeppelin-incubator/bigquery/src/test/java/org/apache/zeppelin/bigquery/BigQueryInterpreterTest.java:[41,41] package com.google.cloud.bigquery.testing does not exist
[ERROR] /Users/..../zeppelin-incubator/bigquery/src/test/java/org/apache/zeppelin/bigquery/BigQueryInterpreterTest.java:[41,41] package com.google.cloud.bigquery.testing does not exist
[INFO] 2 errors
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 20.704 s
[INFO] Finished at: 2016-07-18T11:34:19+09:00
[INFO] Final Memory: 37M/363M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:testCompile (default-testCompile) on project zeppelin-bigquery: Compilation failure: Compilation failure:
[ERROR] /Users/..../zeppelin-incubator/bigquery/src/test/java/org/apache/zeppelin/bigquery/BigQueryInterpreterTest.java:[41,41] package com.google.cloud.bigquery.testing does not exist
[ERROR] /Users/..../zeppelin-incubator/bigquery/src/test/java/org/apache/zeppelin/bigquery/BigQueryInterpreterTest.java:[41,41] package com.google.cloud.bigquery.testing does not exist
[ERROR] -> [Help 1]
```

I'm not 100% sure why is that, may be you could point me to the documentation on some pre-requests that I'm missing?

But in case some useful tests can not work without external service dependency - the simplest way to fix the CI is to:
 - exclude them from the default build and 
 - document the way to run it manually

You can see an example of how the same was recently done for the python interpreter in #1164, in particular regarding [PythonInterpreterWithPythonInstalledTest](https://github.com/apache/zeppelin/pull/1164/files#diff-177dacc81c4b28b15b73263d3540968cL33) and it's excludion in [pom.xml](https://github.com/apache/zeppelin/pull/1164/files#diff-db5740cf5cd90cf11935b92eb80d7908R38)

babupe commented:
Sorry about that, I have removed the bad package import from the class

babupe commented:
Got it, so regarding the copyright license.

Is it ok if I do the following,
1. Create a NOTICE file in the root of bigquery project. i.e ./bigquery/NOTICE and add something like https://github.com/apache/incubator-beam/blob/master/NOTICE
2. Change all copyright headers to standard ASF text.

bzz commented:
Regarding 
>Is it ok if I do the following,
> 1. Create a NOTICE file in the root of bigquery project. i.e ./bigquery/NOTICE and add something like https://github.com/apache/incubator-beam/blob/master/NOTICE

[ASF policy](http://www.apache.org/legal/src-headers.html#notice) requires "Every Apache distribution should include a NOTICE file in the top directory, along with the standard LICENSE file.", so I think we need to keep it in https://github.com/apache/zeppelin/blob/master/NOTICE  

> 2. Change all copyright headers to standard ASF text.

This sounds great

bzz commented:
> Sorry about that, I have removed the bad package import from the class

No problem! Thanks for taking care. According to [CI logs](https://s3.amazonaws.com/archive.travis-ci.org/jobs/145450671/log.txt) there is still some issue with the tests 

```
[INFO] Zeppelin: BigQuery interpreter ..................... FAILURE [  1.003 s]

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.apache.zeppelin.bigquery.BigQueryInterpreterTest
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/home/travis/build/apache/zeppelin/zeppelin-interpreter/target/zeppelin-interpreter-0.7.0-SNAPSHOT.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/travis/build/apache/zeppelin/zeppelin-interpreter/target/lib/slf4j-log4j12-1.7.10.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/travis/.m2/repository/org/slf4j/slf4j-log4j12/1.7.10/slf4j-log4j12-1.7.10.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
log4j:WARN No appenders could be found for logger (org.apache.zeppelin.interpreter.Interpreter).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.3 sec <<< FAILURE! - in org.apache.zeppelin.bigquery.BigQueryInterpreterTest
badSqlSyntaxFails(org.apache.zeppelin.bigquery.BigQueryInterpreterTest)  Time elapsed: 0.156 sec  <<< ERROR!
java.lang.NullPointerException: null
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.run(BigQueryInterpreter.java:263)
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.executeSql(BigQueryInterpreter.java:248)
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.interpret(BigQueryInterpreter.java:289)
	at org.apache.zeppelin.bigquery.BigQueryInterpreterTest.badSqlSyntaxFails(BigQueryInterpreterTest.java:112)

sqlSuccess(org.apache.zeppelin.bigquery.BigQueryInterpreterTest)  Time elapsed: 0 sec  <<< ERROR!
java.lang.NullPointerException: null
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.run(BigQueryInterpreter.java:263)
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.executeSql(BigQueryInterpreter.java:248)
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.interpret(BigQueryInterpreter.java:289)
	at org.apache.zeppelin.bigquery.BigQueryInterpreterTest.sqlSuccess(BigQueryInterpreterTest.java:103)


Results :

Tests in error: 
  BigQueryInterpreterTest.badSqlSyntaxFails:112 Â» NullPointer
  BigQueryInterpreterTest.sqlSuccess:103 Â» NullPointer

Tests run: 2, Failures: 0, Errors: 2, Skipped: 0
```

babupe commented:
I have skipped the failing test and modified licensing and modified the notice at the project root.
Let me know if it looks good.

bzz commented on NOTICE:
```diff
@@ -4,5 +4,6 @@ Copyright 2015 - 2016 The Apache Software Foundation
 This product includes software developed at
 The Apache Software Foundation (http://www.apache.org/).
 
+Portions of this software were developed at Google (http://www.google.com/). Licensed under the Apache v2.0 License.
```
I think we better be more specific here and add pah `./bigquery/` to avoid the confusion.

bzz commented on bigquery/pom.xml:
```diff
@@ -0,0 +1,197 @@
+    <project.http.version>1.21.0</project.http.version>
+    <project.oauth.version>1.21.0</project.oauth.version>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <bigquery.test.exclude>**/BigQueryInterpreterTest.java</bigquery.test.exclude>
```
Excluding tests will fix the CI, but for the future maintainers, assuming they will not know much about it at first, I think we need to have few things documented in `./bigquery/README.md` 
 - a command for manually running it `mvn -Dbigquery.test.exclude='' test -pl bigquery`
 - a pre-requests, that need to be configured (and how) in order for this test to pass locally 

I tried on my local env and got

```
$mvn -Dbigquery.test.exclude='' test -pl bigquery
....
java.lang.NullPointerException: null
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.run(BigQueryInterpreter.java:261)
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.executeSql(BigQueryInterpreter.java:246)
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.interpret(BigQueryInterpreter.java:287)
	at org.apache.zeppelin.bigquery.BigQueryInterpreterTest.badSqlSyntaxFails(BigQueryInterpreterTest.java:110)

sqlSuccess(org.apache.zeppelin.bigquery.BigQueryInterpreterTest)  Time elapsed: 0 sec  <<< ERROR!
java.lang.NullPointerException: null
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.run(BigQueryInterpreter.java:261)
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.executeSql(BigQueryInterpreter.java:246)
	at org.apache.zeppelin.bigquery.BigQueryInterpreter.interpret(BigQueryInterpreter.java:287)
	at org.apache.zeppelin.bigquery.BigQueryInterpreterTest.sqlSuccess(BigQueryInterpreterTest.java:101)

Results :

Tests in error:
  BigQueryInterpreterTest.badSqlSyntaxFails:110 » NullPointer
  BigQueryInterpreterTest.sqlSuccess:101 » NullPointer

Tests run: 2, Failures: 0, Errors: 2, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

which leaves not much clues on what went wrong. What do you think, does it make sence?

bzz replied:
A way to run it is documented under README.md now!

babupe commented:
Yes, the code is failing because it has to call the bigquery external service with a project id that it owns. Currently there is no emulator.
So, you may have to run the tests manually from a GCE instance after setting a project id in resources/constants.json.
After this, running the test with -Dbigquery.test.exclude='' will start passing.
Do you want me to document this in README?

bzz commented on bigquery/src/main/java/org/apache/zeppelin/bigquery/BigQueryInterpreter.java:
```diff
@@ -0,0 +1,329 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
```
I think this license header is still different from a [ASF standard one](http://www.apache.org/legal/src-headers.html#headers) (does not mention NOTICE, have a bit different wording)

bzz commented on bigquery/src/test/java/org/apache/zeppelin/bigquery/BigQueryInterpreterTest.java:
```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
```
I think this license header is still different from a [ASF standard one](http://www.apache.org/legal/src-headers.html#headers) (does not mention NOTICE, have a bit different wording)

bzz commented:
> So, you may have to run the tests manually from a GCE instance after setting a project id in resources/constants.json.
>After this, running the test with -Dbigquery.test.exclude='' will start passing.
Do you want me to document this in README?

Got it, yes, I think this is exactly what needs to be documented, like:
```
Tests XXXX are excluded as they depend on BigQuery service which does not have mock at the moment. So in order to run them manually you need:
 - create a new project in GCE (?with some permissions?)
 -  copy project id from YYYY to `resources/constants.json`
 -  run them from GCE instance using `mvn ....`
```

babupe commented:
Got it, pushed all the changes in now.

babupe commented:
I think excluding constants.json helped.
Let me know if you need anything else from my side.

bzz commented:
Great, thank you! Please, let me make a final pass and I'll get back to you.

bzz commented on bigquery/pom.xml:
```diff
@@ -0,0 +1,197 @@
+
+    <dependency>
+      <groupId>com.mockrunner</groupId>
+      <artifactId>mockrunner-jdbc</artifactId>
```
I'm a bit confused verifying dependency list, could you please point me to the place where this library is used?

AhyoungRyu commented on bigquery/README.md:
```diff
@@ -0,0 +1,105 @@
+
+# Sample Screenshot
+
+![screenshot from 2016-07-12 14 27 30](https://cloud.githubusercontent.com/assets/4242273/16785302/31b104e2-4842-11e6-87c0-b79763dd85c0.png)
```
You can put this screenshot image under `docs/assets/themes/zeppelin/img/docs-img` dir. And can we crop the unnecessary part like this? 
 - Before
![before](https://cloud.githubusercontent.com/assets/10060731/16938815/b5dab9d8-4db6-11e6-9bf6-3220e6afca1d.png)

 - After
<img width="1164" alt="bigquery_result" src="https://cloud.githubusercontent.com/assets/10060731/16938817/b9213ea0-4db6-11e6-8c3b-8149a0bdf874.png">

bzz commented:
Looks great to me, thank you for taking care!

I think now the only thing that is left, is to determine the status of dependency:

```
<groupId>com.google.apis</groupId>
<artifactId>google-api-services-bigquery</artifactId>
<version>v2-rev265-1.21.0</version>
```

>Got it. These packages are licensed under Apache 2.0. I have asked around to see if the code is publicly available.

Is that an open source library? If so, let's link the sources to README.md

babupe commented:
So, I asked around and it looks like these are apiary generated apis and the generation is not external. I was guided to the discovery doc here - https://www.googleapis.com/discovery/v1/apis/bigquery/v2/rest

bzz commented:
I see, generated source code is not available, thank you for clarification! So we should assume that we are including in the Apache Zeppelin release convenience binary a non-opensource third-party binary, distributed under permissive license.

We need a proper attribution then, but to be honest with you -  I think we might need a hand from LEGAL branch of ASF here, as it's the first time we deal with such matter.

I think everything is fine and it should be possible, will just require a propper attribution. I will bet back to you ASAP here, as we triage it with ASF.

bzz commented:
As transitive dependecies in compile scope got included in final release convenience binary, all licenses need to be included to [zeppelin-distribution/src/bin_license/LICENSE](https://github.com/apache/zeppelin/blob/master/zeppelin-distribution/src/bin_license/LICENSE)

I double-checked and have marked the included ones

  - [x] +- com.google.apis:google-api-services-bigquery:jar:v2-rev265-1.21.0:compile
  - [ ] |  \- com.google.api-client:google-api-client:jar:1.21.0:compile
  - [ ] |     \- com.google.guava:guava-jdk5:jar:17.0:compile
  - [ ] +- com.google.oauth-client:google-oauth-client:jar:1.21.0:compile
  - [ ] |  +- com.google.http-client:google-http-client:jar:1.21.0:compile
  - [x] |  |  \- org.apache.httpcomponents:httpclient:jar:4.3.6:compile
  - [x] |  |     +- org.apache.httpcomponents:httpcore:jar:4.3.3:compile
  - [x] |  |     +- commons-logging:commons-logging:jar:1.1.1:compile
  - [x] |  |     \- commons-codec:commons-codec:jar:1.5:compile
  - [x] |  \- com.google.code.findbugs:jsr305:jar:1.3.9:compile
  - [ ] +- com.google.http-client:google-http-client-jackson2:jar:1.21.0:compile
  - [x] |  \- com.fasterxml.jackson.core:jackson-core:jar:2.1.3:compile
  - [ ] +- com.google.oauth-client:google-oauth-client-jetty:jar:1.21.0:compile
  - [ ] |  +- com.google.oauth-client:google-oauth-client-java6:jar:1.21.0:compile
  - [ ] |  \- org.mortbay.jetty:jetty:jar:6.1.26:compile
  - [ ] |     +- org.mortbay.jetty:jetty-util:jar:6.1.26:compile
  - [ ] |     \- org.mortbay.jetty:servlet-api:jar:2.5-20081211:compile
  - [x] +- com.google.code.gson:gson:jar:2.6:compile
  - [x] +- org.slf4j:slf4j-api:jar:1.7.10:compile
  - [x] +- org.slf4j:slf4j-log4j12:jar:1.7.10:compile
  - [x] |  \- log4j:log4j:jar:1.2.17:compile

babupe commented:
No problem. Pushed

bzz commented:
Great, thank you for understanding, we are almost there. I have marked above ones that I was able to verify, could you please double-check it?

On the `com.google.apis:google-api-services-bigquery:v2-rev265-1.21.0` - good news! I managed to find source code of this library, presumably generated from the [endpoint](https://www.googleapis.com/discovery/v1/apis/bigquery/v2/rest) by [google/apis-client-generator](https://github.com/google/apis-client-generator) available under ASL2.0 on Maven Central http://central.maven.org/maven2/com/google/apis/google-api-services-bigquery/v2-rev265-1.21.0/google-api-services-bigquery-v2-rev265-1.21.0-sources.jar

To avoid this confusion further down the road, could you please add those links to `README.md` as part of the interpreter implementation details and, if possible, a brief note why is it preferred over [GoogleCloudPlatform/gcloud-java/#google-cloud-bigquery-alpha](https://github.com/GoogleCloudPlatform/gcloud-java/#google-cloud-bigquery-alpha) ?

babupe commented:
Awesome! Thank you!
I have added that to the documentation.

bzz commented:
Looks great to me, thank you!
I'll do a final pass and play with it more today. 

Will merge it right after, if there is no further discussion.

babupe commented:
Hi Alexander,
I spoke to our OSS team in Google and it seems they wanted to keep the source headers.
Since I have added that to the NOTICE file earlier, I have modified the license headers for the sources to refer to Google license.
Is this OK?

I referred to your earlier comment for this,
"In this case, if you choose to keep it, it makes sense to be more specific and include in NOTICE a location as well: either ./bigquery/ or full paths to the source code files for which Google holds copyright i.e bigquery/src/main/java/org/apache/zeppelin/bigquery/BigQueryInterpreter.java, etc"

bzz commented:
Hi Babupe, in my understanding this change makes ASF treat current patch\PR as a ["third-party work"](http://www.apache.org/legal/src-headers.html#3party) VS ["work submitted directly to the ASF by the copyright owner or owner's agent"](http://www.apache.org/legal/src-headers.html#headers), wich just happen to have Apache 2.0 license.

So according to the policy I reference above
> Do ensure that every third-party work includes its associated license, even if that requires adding a copy of the license from the third-party download site into the distribution.

We need to:
 - add the full text of the license as a separate file under the [./licenses/](https://github.com/apache/zeppelin/tree/master/licenses)
 - remove the line that was added before from NOTICE and 
 - add mention of the `./bigquery` to the Apache2.0 part of the root LICENSE file

Sorry for the bureaucracy again, hope it makes sense to you and please do not hesitate to ask questions in case you have them.

One more thing - there is a trivial conflict with current master now, but it blocks the merge, could you please `git rebase origin master` it?

I plan to merge it to *master*, as well as *branch-0.6* ASAP, so it will become part of the coming 0.6.1 release later this week.

babupe commented:
Made the changes and pushed them. Hope its good now

bzz commented:
Great stuff! Thank you for taking care. 
I have added babupe/zeppelin#1 please review and merge it in.

Will merge this contribution to **master** and **branch-0.6**, if there is no further discussion.

babupe commented:
Awesome! Merged this.

bzz commented:
CI build failed due to networking issues

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-remote-resources-plugin:1.4:process (default) on project zeppelin: Error downloading resources archive. Could not transfer artifact org.apache.apache.resources:apache-jar-resource-bundle:jar:1.5-SNAPSHOT from/to apache-snapshots (https://repository.apache.org/snapshots/): Connect to repository.apache.org:443 [repository.apache.org/207.244.88.143] failed: Connection timed out
[ERROR] org.apache.apache.resources:apache-jar-resource-bundle:jar:1.5-SNAPSHOT
```
which are not related.

bzz commented:
While running locally I got:

```
ERROR [2016-07-29 15:05:09,591] ({pool-2-thread-2} BigQueryInterpreter.java[open]:142) - Cannot open connection
java.io.IOException: The Application Default Credentials are not available. They are available if running in Google Compute Engine. Otherwise, the environment variable GOOGLE_APPLICATION_CREDENTIALS must be defined pointing to a file defining the credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
        at com.google.api.client.googleapis.auth.oauth2.DefaultCredentialProvider.getDefaultCredential(DefaultCredentialProvider.java:95)
        at com.google.api.client.googleapis.auth.oauth2.GoogleCredential.getApplicationDefault(GoogleCredential.java:213)
        at org.apache.zeppelin.bigquery.BigQueryInterpreter.createAuthorizedClient(BigQueryInterpreter.java:155)
        at org.apache.zeppelin.bigquery.BigQueryInterpreter.open(BigQueryInterpreter.java:138)
        at org.apache.zeppelin.interpreter.LazyOpenInterpreter.open(LazyOpenInterpreter.java:69)
        at org.apache.zeppelin.interpreter.LazyOpenInterpreter.interpret(LazyOpenInterpreter.java:93)
        at org.apache.zeppelin.interpreter.remote.RemoteInterpreterServer$InterpretJob.jobRun(RemoteInterpreterServer.java:383)
        at org.apache.zeppelin.scheduler.Job.run(Job.java:176)
        at org.apache.zeppelin.scheduler.FIFOScheduler$1.run(FIFOScheduler.java:139)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:744)
 INFO [2016-07-29 15:05:09,593] ({pool-2-thread-2} BigQueryInterpreter.java[close]:281) - Close bqsql connection!
```

@babupe do you think it would make sense to expose this though interpreter configuration properties, same as `project id`, so people could override it? Or may be it would be easier to use OAuth service account?

In any way, we need to document this properly for people to know that there are [pre-requests](https://developers.google.com/identity/protocols/application-default-credentials), before using this interpreter.

I will submit the PR with docs

bzz commented:
Added babupe/zeppelin#2 

After configuring credentials I got strange behavior: query `show tables` results is `ERROR` paragraph status with 
 - no output in GUI
 - and nothing in the logs

![screen shot 2016-07-29 at 15 40 08](https://cloud.githubusercontent.com/assets/5582506/17239974/3869e042-55a3-11e6-9861-ecc4cc9f6247.png)

```
  INFO [2016-07-29 15:39:49,759] ({pool-2-thread-2} SchedulerFactory.java[jobStarted]:131) - Job remoteInterpretJob_1469774389758 started by scheduler org.apache.zeppelin.bigquery.BigQueryInterpreter745463803
 INFO [2016-07-29 15:39:49,759] ({pool-2-thread-2} BigQueryInterpreter.java[interpret]:288) - Run SQL command 'show tables'
 INFO [2016-07-29 15:39:50,531] ({pool-2-thread-2} SchedulerFactory.java[jobFinished]:137) - Job remoteInterpretJob_1469774389758 finished by scheduler org.apache.zeppelin.bigquery.BigQueryInterpreter745463803
```

Which is a bit frustrating experience, since it is `ERROR` but user does not know why.

babupe commented:
Sorry about the delay. Merged it now. Thank you so much for documenting this!
I am looking at surfacing the errors out to the interpreter. Will have an update on it.
Will also look at exposing creds for non GCE instances through configuration.

bzz commented:
Thank you! Sounds awesome to me.

babupe commented:
I have now pushed the change to capture and show bigquery exceptions with bad statements both in logs and interpreter. Will look into the credentials.

babupe commented:
Do you think the external auth could be a blocker for the initial merge?
If not, lets table it for now and I will pick this up later?
It may be a little bit more involved than I thought.

bzz commented:
Got it, makes perfect sense! Thank you for updating error handling and the plan to address external auth cfg later sounds good.

Looks great to me, merging to *master* and *branch-0.6* if there is not further duscussion!
